syntax = "proto3";

package rooms_service;

// Il servizio deve permettere al client di:
// 1. ricevere aggiornamenti in real-time delle stanze
// 2. creare una nuova stanza
// 3. entrare/uscire da una stanza
// 4. conoscere il numero di partecipanti di una stanza

service RoomsServiceInterface
{
	// Stream continuo di aggiornamenti in real-time delle stanze.
	// rpc WatchRoomsProcedure(WatchRoomsRequest) returns (stream RoomList) {}
 	// Il client manda una singola richiesta al server per ottenere la lista di tutte le stanze esistenti
 	// rpc ListAllRoomsProcedure(ListAllRoomsRequest) returns (RoomList) {}
	
  // Crea una nuova stanza
	rpc CreateRoomProcedure(CreateRoomRequest) returns (CreateRoomResponse) {}
	// Elimina una stanza
	rpc DeleteRoomProcedure(DeleteRoomRequest) returns (DeleteRoomResponse) {}
   
  
  // Entra in una stanza
  // rpc JoinRoom(JoinRoomRequest) returns (JoinRoomResponse) {}
  // Esce da una stanza
  // rpc LeaveRoom(LeaveRoomRequest) returns (LeaveRoomResponse) {}
}

message WatchRoomsRequest 
{
  uint32 client_id = 1;
}
message RoomInfo
{
	uint32 room_id = 1;
	uint32 creator_id = 2;
	string room_name = 3;
	uint32 user_count = 4;
}
message RoomList 
{
  repeated RoomInfo rooms = 1;
}
message ListAllRoomsRequest {}

message CreateRoomRequest
{
	uint32 creator_id = 1;
  string room_name = 2;
}
message CreateRoomResponse
{
  RoomInfo room = 1;
}

message DeleteRoomRequest
{
	uint32 room_id = 1;
	uint32 client_id = 2;
}
message DeleteRoomResponse{}


// 
// message JoinRoomRequest
// {
//   string client_id = 1;
//   string room_id = 2;
// }
// 
// message JoinRoomResponse
// {
//   bool success = 1;
//   string error = 2;
// }
// 
// message LeaveRoomRequest
// {
//   string client_id = 1;
//   string room_id = 2;
// }
// 
// message LeaveRoomResponse 
// {
//   bool success = 1;
//   string error = 2;
// }